Cumulative Governor changes and L2 canceller role

corydickson opened the pull request:
This PR combines the changes to the governor to support the new proposal creation role as well as the scripts to add new canceller roles to the timelock. 

Note: During the upgrade be sure to deploy a timelock with `AccessControlEnumerable` to ensure complete functionality of the scripts.

kent committed:
added L2 canceller safe role script

kent committed:
added more robust simulation and verfiication with pretty human readable

kent committed:
added in bash scripts

kent committed:
updated the Readme

0xOneTony committed:
feat: add proposer role

0xOneTony committed:
fix: readme

0xOneTony committed:
fix: tests

0xOneTony committed:
fix: warning

0xOneTony committed:
fix: grammar

0xOneTony committed:
chore: change test name

corydickson committed:
feat: add proposer role (#43)

Adds support for a new `authorizedProposer` role. This role will be able
to create proposals along with the manager and timelock. Initially will
be set to the same address as the `manager`, however eventually it will
change using the `setAuthorizedProposer` to the new `ProposalValidator`
contract when it is ready.

0xOneTony commented on script/SetCancellers.s.sol:
```diff
@@ -0,0 +1,108 @@
+            console.log("L2 Safe 3 already has CANCELLER_ROLE:", L2_SAFE_3);
+        }
+
+        // Step 2: Ensure all proposers have CANCELLER_ROLE (which is the default in OpenZeppelin's TimelockController)
```
is there a specific reason you check for all proposers and not just the 3 Safe addresses?

kent replied:
Just for completion sake. Do you see a harm in being verbose here, or just asking?

0xOneTony commented on script/SetCancellers.s.sol:
```diff
@@ -0,0 +1,108 @@
+        console.log("L2 Safe 2 has CANCELLER_ROLE:", timelock.hasRole(CANCELLER_ROLE, L2_SAFE_2));
+        console.log("L2 Safe 3 has CANCELLER_ROLE:", timelock.hasRole(CANCELLER_ROLE, L2_SAFE_3));
+
+        // Verify all proposers have the CANCELLER_ROLE
```
Same here why we check for all addresses? Wouldnt it make sense to check if the 3 safes also have/dont have the "Proposer" role?

kent committed:
fix: reduce upper bound of fuzz testing on warp

kent committed:
fix: build warnings with view

kent committed:
feat: add RedeployTimelock script to deploy with cancellers

kent committed:
fix: ignore size for tests in CI

corydickson committed:
Change timelock initialization with proxy pattern

corydickson committed:
Remove updateTimelock step when redeploying timelock

corydickson committed:
Correct the documentation for propose and prevent the proxy admin from proposing during fuzz test

0xOneTony committed:
fix: bump governor version

corydickson committed:
Revert if authorized proposer is not different

corydickson committed:
Initialize timelock in a single transaction

corydickson committed:
Draft: show revert without condition

0xOneTony committed:
fix: remove timelock as proposer

corydickson committed:
Remove duplicate grant roles

corydickson committed:
Add indexed parameter

corydickson committed:
Remove manager from timelock script

corydickson committed:
Add correct fuzzing check to show revert when timelock address is zero

corydickson committed:
L-01 Deployment Script Incorrectly Calls updateTimelock Function (#47)

corydickson committed:
L-02 Incorrect Documentation &  L-03 Incomplete Documentation (#48)

Took this opportunity to fix a flaking test when the fuzzer impersonates
the proxy admin address and causes a failure on the
TransparentUpgradeableProxy for the governor.

corydickson committed:
L-06 Bump Governor Version (#49)

Per OZ's audit find L-06, this PR bumps the version of the Governor
contract.

> The OptimismGovernor contract was modified to include new logic for
the authorized proposer. However, the VERSION was not incremented. As a
result, the reinitialize function is not callable and will revert due to
the reinitializer modifier. While it is possible to set the
authorizedProposer using the setAuthorizedProposer function, this
approach bypasses the standard practice of handling state changes during
upgrades. Consider increasing the VERSION of the contract and setting
the parameters through the reinitialize logic.

corydickson committed:
N-04 Inconsistency in Indexing Event Parameters (#51)

corydickson committed:
N-02 Repeated Calls to Grant Role (#52)

corydickson committed:
L-04 Possible Duplicate Event Emissions (#53)

corydickson committed:
L-05 Timelock Proxy Deployment and Initialization Split into Separate Transactions (#54)

corydickson committed:
Merge branch 'op-gov-changes' into N-01

corydickson committed:
N-01 Discrepancies between Documentation and Implementation (#55)

corydickson committed:
L-07 updateTimelock Lacks Zero-Address Validation (#56)

0xOneTony committed:
Merge branch 'op-gov-changes' of https://github.com/voteagora/optimism-governor into fix/l08-remove-unused-check

corydickson committed:
L-08 Remove timelock as proposer (#50)

Per OZ's audit report, this PR addresses the removal of the `timelock`
as a proposer, making only the `manager` and `authroizedProposer` as
valid proposers.

> The proposeWithModule function is marked with the onlyValidProposer
modifier, which allows the function to be executed by one of the
following addresses: manager , authorizedProposer , or timelock .
However, within the function, there is a check that requires the
following: if the function is called by an address that is neither
manager nor authorizedProposer , a proposal votes threshold needs to be
met. Given that the only other address allowed to call the
proposeWithModule function is timelock , the aforementioned check
implies that timelock itself must meet a vote threshold. This design is
inconsistent with the propose function where the check for privileged
proposer addresses to meet the proposal threshold is intentionally
removed. In addition, this discrepancy raises a further discussion on
whether the timelock contract should be allowed to be a proposer on the
governor. Timelocks are typically meant to delay, queue, and execute
proposals. Allowing the timelock contract to propose can create circular
governance paths, which can prove to be error-prone. Consider unifying
the behavior for the proposal-threshold check across the propose and
proposeWithModule functions. Moreover, reconsider the design decision of
allowing the timelock contract to be a proposer.
